trie/zkproof/writer.go

```diff
@@ -0,0 +1,729 @@
+	zkDb := trie.NewZktrieDatabase(underlayerDb)
+
+	accounts := make(map[common.Address]*types.StateAccount)
+
+	// resuming proof bytes to underlayerDb
+	for addrs, proof := range storage.Proofs {
+		if n := resumeProofs(proof, underlayerDb); n != nil {
+			addr := common.HexToAddress(addrs)
+			if n.Type == trie.NodeTypeEmpty {
+				accounts[addr] = nil
+			} else if acc, err := types.UnmarshalStateAccount(n.Data()); err == nil {
+				if bytes.Equal(n.NodeKey[:], addressToKey(addr)[:]) {
+					accounts[addr] = acc
+				} else {
+					// should still mark the address as being trace (data not existed yet)
+					accounts[addr] = nil
+				}
+
+			} else {
+				return nil, fmt.Errorf("decode account bytes fail: %s, raw data [%x]", err, n.Data())
+			}
+
+		} else {
+			return nil, fmt.Errorf("can not resume proof for address %s", addrs)
+		}
+	}
+
+	storages := make(map[common.Address]*trie.ZkTrie)
+
+	for addrs, stgLists := range storage.StorageProofs {
+
+		addr := common.HexToAddress(addrs)
+		accState, existed := accounts[addr]
+		if !existed {
+			// trace is malformed but currently we just warn about that
+			log.Warn("no account state found for this addr, mal records", "address", addrs)
+			continue
+		} else if accState == nil {
+			// create an empty zktrie for uninit address
+			storages[addr], _ = trie.NewZkTrie(common.Hash{}, zkDb)
+			continue
+		}
+
+		for keys, proof := range stgLists {
+
+			if n := resumeProofs(proof, underlayerDb); n != nil {
+				var err error
+				storages[addr], err = trie.NewZkTrie(accState.Root, zkDb)
+				if err != nil {
+					return nil, fmt.Errorf("zktrie create failure for storage in addr <%s>: %s", err, addrs)
+				}
+
+			} else {
+				return nil, fmt.Errorf("can not resume proof for storage %s@%s", keys, addrs)
+			}
+
+		}
+	}
+
+	zktrie, err := trie.NewZkTrie(
+		storage.RootBefore,
+		trie.NewZktrieDatabase(underlayerDb),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("zktrie create failure: %s", err)
+	}
+
+	// sanity check
+	if !bytes.Equal(zktrie.Hash().Bytes(), storage.RootBefore.Bytes()) {
+		return nil, fmt.Errorf("unmatch init trie hash: expected %x but has %x", storage.RootBefore.Bytes(), zktrie.Hash().Bytes())
+	}
+
+	return &zktrieProofWriter{
+		db:                  zkDb,
+		tracingZktrie:       zktrie,
+		tracingAccounts:     accounts,
+		tracingStorageTries: storages,
+	}, nil
+}
+
+const (
+	posSSTOREBefore = 0
+	posCREATE       = 0
+	posCREATEAfter  = 1
+	posCALL         = 2
+	posSTATICCALL   = 0
+)
+
+func getAccountState(l *types.StructLogRes, pos int) *types.AccountWrapper {
+	if exData := l.ExtraData; exData == nil {
+		return nil
+	} else if len(exData.StateList) < pos {
+		return nil
+	} else {
+		return exData.StateList[pos]
+	}
+}
+
+func copyAccountState(st *types.AccountWrapper) *types.AccountWrapper {
+	return &types.AccountWrapper{
+		Nonce:    st.Nonce,
+		Balance:  st.Balance,
+		CodeHash: st.CodeHash,
+		Address:  st.Address,
+		Storage:  st.Storage,
```

Comment:
can it really copy the AccountState instead of re-use the pointers of `Balance` & `Storage`?

Reply:
So you prefer to we should have a deep copy of these 2 fields？

Reply:
If using the pointer is fine then it's fine.

TBH I don't know what is desired.